Project includes request

srallen opened the pull request:
Closes #47. This is a branch off of #45, so that has to merge first and then I'll rebase and change the base to master. ~WIP because I need to update docs, but the code and tests are all done.~

This adds a helper for project request including avatar, background, owner, and pages. This adds resource mocks for those and includes a fairly substance refactor on those mocks (we haven't published yet, so I'm not worried about semvar yet). I'm happy to update any tests in app-project that may be using them, but I think this refactor makes usage more clear and maintainable. The variable naming before was unwieldy. So now if you want the project endpoint or the mocks, it's like this:
```
import { projects } from '@zooniverse/lib-panoptes-js`

const mockProjectResource = projects.mocks.resources.projectOne
const mockProjectGetResponse = projects.mocks.responses.get.project
```

Requests are still the same: 
```
import { projects } from '@zooniverse/lib-panoptes-js`
\\ should return the first page of projects
projects.get()
```

I also broke apart the projects.js into two separate files, one for the standard REST requests and one for the common requests. I think this may set us up to be able to more easily add helpers that use graphql.

srallen commented:
@eatyourgreens yes, we should start to think about that. I don't see translations as an include on the project resource serializer (https://github.com/zooniverse/Panoptes/blob/master/app/serializers/project_serializer.rb#L33-L34), and I'm not sure if it should be. I could open another PR to add basic translation resource CRUD requests? Did you have something else in mind? FWIW, I only want this lib to handle API requests, not app state at all.

eatyourgreens commented:
In PFE, at the moment, translations are loaded in via separate API calls eg. https://github.com/zooniverse/Panoptes-Front-End/blob/eef8e357d6dea221d49adaa235c94dc79a2f149f/app/pages/project/index.jsx#L207

I don't think you can include them in the project request. Even so, you'd still need to request them again if the end user changed the language.

There's a redux action defined in PFE for loading translations, which might be helpful. In the new classifier, we can probably render the translation keys directly, wherever text is printed, rather than trying to copy the structure of the resource objects like PFE does (for backwards compatibility with older rendering code.)

eatyourgreens commented:
On the app state side, we should localise as early as possible in the implementation. Going through PFE to retrospectively localise hardcoded strings was a pain.

srallen commented:
I've added these comments to a separate issue so we can keep track after this PR closes.

srallen commented:
Docs are done.

srallen commented:
I rebased, though, there are now two failing test. I might've rebase one conflict incorrectly. Working on fixing these....

srallen commented:
Not a rebase issue, but rather the default params are now being used so the tests were correctly now failing.

srallen commented:
Rebased again.

srallen commented:
Ok, time to rebase this one.

srallen commented:
Ok, rebased and ready for review again.

eatyourgreens commented on packages/lib-panoptes-js/src/docs/media.md:
```diff
@@ -0,0 +1,55 @@
+  content_type: "image/jpeg",
+  created_at: "2016-12-07T23:11:47.452Z",
+  external_link: false,
+  href: '/users/24/avatar`,
```
backtick instead of single quote here?

eatyourgreens commented on packages/lib-panoptes-js/src/docs/media.md:
```diff
@@ -0,0 +1,55 @@
+  created_at: "2016-12-07T23:11:47.452Z",
+  external_link: false,
+  href: '/users/24/avatar`,
+  id: '24,
```
Typo: missing quote mark.

Should strings be doubled-quoted or single-quoted? The example mixes both.

srallen replied:
I've updated to be single-quoted to be consistent with what's in the javascript. This was just me being sloppy with copy and past I think 😬

eatyourgreens commented on packages/lib-panoptes-js/src/docs/media.md:
```diff
@@ -0,0 +1,55 @@
+  links: {},
+  media_type: 'users_avatar',
+  metadata: null,
+  src: `https://panoptes-uploads.zooniverse.org/production/users_avatar/random-string-example.jpeg`
```
Typo: backticks should be quotes here?

eatyourgreens commented on packages/lib-panoptes-js/src/resources/projects/rest.js:
```diff
@@ -0,0 +1,41 @@
+
+  if (!projectId) return panoptes.get(endpoint, queryParams)
+  if (projectId && typeof projectId !== 'string') return handleError('Projects: Get request id must be a string.')
+
```
Should this be `handleError(new Error('Projects: Get request id must be a string.')`?

srallen replied:
Thanks for the examples. I'll update this.

eatyourgreens commented on packages/lib-panoptes-js/src/resources/projects/rest.js:
```diff
@@ -0,0 +1,41 @@
+  const { id, data } = params
+  if (id && typeof id !== 'string') return handleError('Projects: Update request id must be a string.')
+  if (id && data) return panoptes.put(`${endpoint}/${id}`, data)
+  if (!id) return handleError('Projects: Update request missing project id.')
```
Should this string be an Error too?

rogerhutchings commented:
MDN reckons rejecting a promise with an Error object is best practice as it provides a stack trace: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/reject

srallen commented:
@eatyourgreens Ok, I think I've fixed everything from your review. As a side note, I moved the docs folder to where it should be so github correctly picks up the template files in there.

eatyourgreens commented on packages/lib-panoptes-js/src/panoptes.js:
```diff
@@ -3,14 +3,14 @@
 const { config } = require('./config')
 
-function handleErrors (error) {
+function handleError (error) {
```
There's also a function called `handleError` in `resources/projects/helpers.js`, which could be confusing. I was looking at the helper function when I commented yesterday.

The helper raises an error, rather than handling it, so maybe rename that one to `raiseError`?

This function handles errors, so it should expect to receive an Error and, I think, return something that looks like an API response (an empty array or null maybe?) Reason for this is that the return value of a catch block is passed into the next then block in a promise chain. At the moment, any function chained after the catch would receive an error as the input, and I'm not sure that this is what the code is intended to do.

srallen replied:
The `handleError` here is just making sure the error gets logged to the console if the console is available, then passing it along. I can remove it to wholly put the responsibility of handling the superagent request errors on to the consumer of this library.

eatyourgreens replied:
Ok. If you want to pass the error back to the consumer, I think you want to rethrow here, after logging to the console. Otherwise, the catch block will have caught the error and the promise chain will continue as if the error has been handled.

srallen replied:
I'm thinking through this and I'm remembering now that there may be some cases where we don't want the error to be logged like the 404s for projects or users that don't have an avatar. I think I'll just remove it.

srallen commented:
I've done bit of reorganizing and renaming again. `handleError` in the projects folder is now `raiseError` and is in the index file of the now renamed `utilityFunctions` folder for use by the whole library. `raiseError` now has a `errorClass` parameter and will create an error of the correct class type when defined. 

`handleError` that was in the panoptes.js file is now removed. The consumer of the library should decide how to catch and handle errors and if they want the error logged.

srallen commented:
I'm going to merge, so if there are any further comments, please open an issue.

eatyourgreens commented:
Looks good to me. 👍
